text/0030-link-system.md

```diff
@@ -0,0 +1,148 @@
+# Link System
+- Start Date: 2021/10/21
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/30
+
+## Decision, including impact on distributions
+
+We will extend the extension system to have support specific to links.
+
+`esm-framework` should define the following function:
+
+```ts
+function createLinkExtension(params: {
+  label: string | () => string | () => Promise<string>,
+  to?: string | () => string | () => Promise<string>,
+  onClick?: (event, props) => void
+}): () => Promise<ReactAppOrParcel<any>>
+```
+
+Where `to` can include `${variables}` which will be interpolated from
+`openmrsBase`, `openmrsSpaBase`, or the extension props; and where
+`onClick` receives extension props as its second argument.
+
+This can be used in an extension definition, for example:
+
+```ts
+function setupOpenMRS() {
+  return {
+    pages: [],
+    extensions: [{
+      id: 'Patient list link',
+      slot: 'App menu',
+      load: createLinkExtension({
+        label: () => t("patientList", "Patient List"),
+        to: "${openmrsSpaBase}/patient-list/${patientUuid}",
+        onClick: (e, { patientUuid }) => {
+          console.log(`Clicked ${e.button} for patient ${patientUuid}`);
+        }
+      })
+    }]
+  }
+}
+```
+
+In cases where the extension `meta` property is not required, the
+following could equivalently be used:
+
+```ts
+function setupOpenMRS() {
+  return {
+    pages: [],
+    extensions: [],
+    links: [{
+      id: "Patient list link", 
+      label: () => t("patientList", "Patient List"),
+      to: "${openmrsSpaBase}/patient-list",
+      onClick: (e) => { console.log("Clicked button " + e.button); }
+      menu: 'App menu',
+    }]
+  }
+}
+```
+
+We will define a new React component, `Menu` (along with
+framework-independent functions that would support it). `Menu` would
+work exactly like `ExtensionSlot`, but only for Links.
+
+```ts
+function Menu(props: { name: string, state?: object });
+```
+
+We will extend the configuration system so that every Menu supports
+the same configuration options as ExtensionSlots.
+
+The `configure` value will support only the key `openInNewTab`.
+`openInNewTab` will default to `true` for external links, `false` otherwise.
+When `true`, it will add `target="_blank"` to the link. If the link is an
+external link, it will also add `rel="noopener noreferrer"`.
+
+Menu configuration will also support one additional key, `links`, which would
+allow an implementer to add arbitrary links to the Menu:
+
+```json
+{
+  "@openmrs/esm-primary-navigation-app": {
+    "menus": {
+      "App menu": {
+        "add": ["Patient list link"],
+        "remove": ["Provider management link"],
+        "order": ["Home page link", "Patient list link"],
+        "links": [{
+          "label": "Home",
+          "to": "${openmrsSpaBase}/home"
+        }],
+        "configure": {
+          "Patient list link": {
+            "openInNewTab": true
+          }
+        }
+      }
+    }
+  }
+}
+```
+
```

Comment:
@jonathandick @mks-d @manuelroemer

Reply:
My understanding right now is that links _do_ appear in the patient chart based on whether you are online or offline though I don't know how this is done. @manuelroemer could you comment?

Reply:
@jonathandick Yes, that's true! The patient chart is a good example for a menu where some links aren't available when offline. This behavior must be possible.
I want to remark that extensions *are not* available offline unless they opt into offline mode (i.e. `offline` is `false` by default unless specifically configured otherwise). This means that we require the opposite behavior: We want links to be able to specifiy that they *should* render in offline mode.

However, given that the links are still normal extensions (albeit abstracted away), it should already be possible to specify the `online`/`offline` configurations with the registration proposed by @brandones:
```diff
function setupOpenMRS() {
  return {
    pages: [],
    extensions: [{
      id: 'Patient list link',
      slot: 'App menu',
      load: createLinkExtension({
        label: () => t("patientList", "Patient List"),
        to: "${openmrsSpaBase}/patient-list/${patientUuid}",
        onClick: (e, { patientUuid }) => {
          console.log(`Clicked ${e.button} for patient ${patientUuid}`);
        }
      }),
+     online: true,
+     offline: true
    }]
  }
}
```